### PR TITLE
XrCamera: use ppx::Camera instead of custom methods.

### PR DESCRIPTION
**Pull request by hysw**

### CODE REVIEW

**Comment by hysw** [comment]
Given we have a `Camera` interface, have a separate `GetViewMatrixForCurrentView`/`GetProjectionMatrixForCurrentViewAndSetFrustumPlanes` feels odd.

Note, this is stacked on top of #388 due to them having merge conflict.

File: include/ppx/imgui_impl.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ class ImGuiImpl
     virtual void   Shutdown(ppx::Application* pApp) = 0;
     virtual void   NewFrame();
     virtual void   Render(grfx::CommandBuffer* pCommandBuffer) = 0;
+    virtual void   ProcessEvent() {}
```
Comment (wangra-google) [inline include/ppx/imgui_impl.h:40]: this file belongs to a different PR?

File: benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp [hunk 2]
```diff
@@ -1074,15 +1074,13 @@ void GraphicsBenchmarkApp::Render()
     // Reset query
     frame.timestampQuery->Reset(/* firstQuery= */ 0, frame.timestampQuery->GetCount());
 
-    // Update scene data
-    frame.sceneData.viewProjectionMatrix = mCamera.GetViewProjectionMatrix();
 #if defined(PPX_BUILD_XR)
```
Comment (wangra-google) [inline benchmarks/graphics_pipeline/GraphicsBenchmarkApp.cpp:1077]: probably better as a separate function in case other part of the code need to access the camera

File: projects/fishtornado_xr/FishTornado.cpp [hunk 3]
```diff
@@ -493,13 +493,14 @@ void FishTornadoApp::UpdateTime()
 void FishTornadoApp::UpdateScene(uint32_t frameIndex)
 {
     PerFrame& frame = mPerFrame[frameIndex];
-
```
Comment (wangra-google) [inline projects/fishtornado_xr/FishTornado.cpp:496]: probably better as a separate function in case other part of the code need to access the camera

File: src/ppx/imgui_impl.cpp [hunk 4]
```diff
@@ -374,6 +374,51 @@ void ImGuiImplVk::Render(grfx::CommandBuffer* pCommandBuffer)
     ImGui::Render();
     ImGui_ImplVulkan_RenderDrawData(ImGui::GetDrawData(), grfx::vk::ToApi(pCommandBuffer)->GetVkCommandBuffer());
 }
+
```
Comment (wangra-google) [inline src/ppx/imgui_impl.cpp:377]: part of the other PR?

File: src/ppx/xr_component.cpp [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 #if defined(PPX_BUILD_XR)
```
Comment (wangra-google) [inline src/ppx/xr_component.cpp:15]: this file is part of the other PR?

File: include/ppx/xr_component.h [hunk 6]
```diff
@@ -168,15 +192,20 @@ class XrComponent
     XrSession  GetSession() const { return mSession; }
     void       SetCurrentViewIndex(uint32_t index) { mCurrentViewIndex = index; }
     uint32_t   GetCurrentViewIndex() const { return mCurrentViewIndex; }
+    XrCamera&  GetCamera();
```
Comment (wangra-google) [inline include/ppx/xr_component.h:195]: const XrCamera& GetCamera() const;?

File: src/ppx/xr_component.cpp [hunk 7]
```diff
@@ -663,69 +831,52 @@ bool XrComponent::RemoveLayer(LayerRef layerRef)
...
+        mProjectionMatrix = glm::frustumRH_ZO(left, right, down, up, mNearClip, mFarClip);
+    }
+    // Calculate view
+    {
+        const quat orientation = FromXr(view.pose.orientation);
 
-    return glm::make_mat4(mat);
-}
+        mEyePosition   = FromXr(view.pose.position);
+        mViewDirection = glm::rotate(orientation, float3(0, 0, -1));
```
Comment (wangra-google) [inline src/ppx/xr_component.cpp:874]: out of curiosity, why is float3(0, 0, -1)?

File: include/ppx/imgui_impl.h (hunk 1 above)
Comment (hysw) [reply include/ppx/imgui_impl.h:40]: Yes, only as commented above, this is stacked on top of #388 only the last commit is for this PR.

File: include/ppx/xr_component.h (hunk 6 above)
Comment (hysw) [reply include/ppx/xr_component.h:195]: Yes, that's should be in the cleaned up version, this is a draft and I'm still going through the detail of controller input stuff.
`UpdateCamera` probably belong to `BeginFrame` and this function can be constant.

File: src/ppx/xr_component.cpp (hunk 7 above)
Comment (hysw) [reply src/ppx/xr_component.cpp:874]: `-Z` is OpenXR forward direction.

**Review by wangra-google — APPROVED** [review]

File: include/ppx/xr_component.h [hunk 8]
```diff
@@ -95,6 +96,21 @@ struct XrComponentResolution
     uint32_t height = 0;
 };
 
+class XrCamera : public Camera
+{
+public:
+    XrCamera() = default;
+
+    ppx::CameraType GetCameraType() const final { return ppx::CAMERA_TYPE_UNKNOWN; }
```
Comment (apazylbe) [inline include/ppx/xr_component.h:104]: Should this be CAMERA_TYPE_PERSPECTIVE or a new type?

File: include/ppx/xr_component.h [hunk 9]
```diff
@@ -240,6 +260,7 @@ class XrComponent
     std::vector<XrView>                  mViews;
     std::vector<XrEnvironmentBlendMode>  mBlendModes;
     uint32_t                             mCurrentViewIndex = 0;
+    std::vector<XrCamera>                mCameras;
```
Comment (apazylbe) [inline include/ppx/xr_component.h:263]: Can you add a comment to note that there is a camera for each view?

**Review by apazylbe — APPROVED** [review]

File: src/ppx/xr_component.cpp [hunk 10]
```diff
@@ -856,40 +847,72 @@ glm::mat4 XrComponent::GetProjectionMatrixForCurrentViewAndSetFrustumPlanes(floa
...
+    {
+        const XrFovf& fov = mView.fov;
+
+        const float left  = tanf(fov.angleLeft) * mNearClip;
+        const float right = tanf(fov.angleRight) * mNearClip;
+
+        const float down = tanf(fov.angleDown) * mNearClip;
+        const float up   = tanf(fov.angleUp) * mNearClip;
+
+        // Not sure why it is frustumRH_ZO
```
[the remark this answers is not in the document]
Comment (hysw) [reply src/ppx/xr_component.cpp:902]: I'm still trying to figure out if `glm::frustum` would work or do I need this specific variant, it was hardcoded matrix coefficient.

File: projects/04_cube_xr/CubeXr.cpp [hunk 11]
```diff
@@ -275,8 +275,10 @@ void CubeXrApp::Render()
         float4x4 V = glm::lookAt(float3(0, 0, 0), float3(0, 0, 1), float3(0, 1, 0));
 
         if (IsXrEnabled()) {
-            P = GetXrComponent().GetProjectionMatrixForCurrentViewAndSetFrustumPlanes(0.001f, 10000.0f);
-            V = GetXrComponent().GetViewMatrixForCurrentView();
+            GetXrComponent().SetFrustumPlanes(0.001f, 10000.0f);
```
[the remark this answers is not in the document]
Comment (hysw) [reply projects/04_cube_xr/CubeXr.cpp:278]: I don't know why it is this specific value, `PPX_CAMERA_DEFAULT_{NEAR,FAR}_CLIP = {0.1f, 10000.0f}` which is different.

File: include/ppx/xr_component.h (hunk 8 above)
Comment (hysw) [reply include/ppx/xr_component.h:104]: It's a perspective projection, however it is different from `PerspCamera`, as far as I can see this attribute of camera is unused so I put it as `CAMERA_TYPE_UNKNOWN` and we can figure out what exactly it should be when we need the camera type.

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply projects/04_cube_xr/CubeXr.cpp:278]: Does it changes something to push the near plane a bit further?
I suppose this could be the same for both cameras. Maybe 10cm was considered too close when trying XR, so it was reduced to 1mm?
I'd assume it would be safe to set it to `0.1f` for both camera types.

File: src/ppx/xr_component.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply src/ppx/xr_component.cpp:902]: My understanding is as long as we don't define `GLM_FORCE_LEFT_HANDED` (which doesn't seem the case), GLM defaults to `RH` so this part shouldn't change.
However, since we use [0;1] for depth, the `math_config.h` defines GLM_FORCE_DEPTH_ZERO_TO_ONE, so I'd say in the current config, those are equivalent:

```
glm::frustum
glm::frustumZO
glm::frustumRH_ZO
```

I'd be in favor of using `glm::frustum` as it would allow changing the defines in `ppx/math_config.h` and keep the correct behavior, with the caveats that we should define the unit vectors `up`/`right`/`forward` and the `near`/`far` distances someplace unique.

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (wangra-google) [reply projects/04_cube_xr/CubeXr.cpp:278]: should be fine to use the same camera setup for now.

File: src/ppx/xr_component.cpp [hunk 12]
```diff
@@ -856,40 +847,72 @@ glm::mat4 XrComponent::GetProjectionMatrixForCurrentViewAndSetFrustumPlanes(floa
...
+    mFarClip  = farZ;
+    UpdateCamera();
 }
 
-XrPosef XrComponent::GetPoseForCurrentView() const
+void XrCamera::UpdateCamera()
 {
-    PPX_ASSERT_MSG((mCurrentViewIndex < mViews.size()), "Invalid view index!");
-    return mViews[mCurrentViewIndex].pose;
+    const float3 forward = float3(0, 0, -1);
```
[the remark this answers is not in the document]
Comment (wangra-google) [reply src/ppx/xr_component.cpp:877]: better to be a local const, camera forward could be different depending on how the camera is setup.

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply projects/04_cube_xr/CubeXr.cpp:278]: Sorry I meant to say `0.001` for both.

File: src/ppx/xr_component.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply src/ppx/xr_component.cpp:877]: If that's the case, then we shouldn't use `glm::frustum` as discussed below, as we don't assume RH coordinates for the whole BW project (unless ZERO_TO_ONE depth as marked in `ppx/math_config`)

[the remark this answers is not in the document]
Comment (Keenuts) [reply src/ppx/xr_component.cpp:877]: So I'd even suggest if we want to allow FWD/RIGHT/UP to change depending on the camera, that we might want a `getForward()`, `getUp()` and `getRight()` functions in the camera implem that returns those unit vectors

[the remark this answers is not in the document]
Comment (wangra-google) [reply src/ppx/xr_component.cpp:877]: i don't think it is necessary though, in majority of the cases, those 3 vectors are only needed to generate view matrix. once we have the view matrix, the user could retrieve whatever vector is needed.

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (wangra-google) [reply projects/04_cube_xr/CubeXr.cpp:278]: we could use the same near&far to calculate the projection matrix for both xr and non-xr cases.

File: src/ppx/xr_component.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (hysw) [reply src/ppx/xr_component.cpp]: There's the definition in `camera.h`, so it should be fine for now. If that's gets changed we might need to fix something (although I don't expect it to change).

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (hysw) [reply projects/04_cube_xr/CubeXr.cpp:278]: From my perspective, those values are arbitrary, so I could change it to something in this pr if we decided on some value or they can be updated later.

[the remark this answers is not in the document]
Comment (Keenuts) [reply projects/04_cube_xr/CubeXr.cpp:278]: I'd be in favor of using `PPX_CAMERA_DEFAULT_{NEAR,FAR}_CLIP` then

File: src/ppx/xr_component.cpp [hunk 13]
```diff
@@ -856,40 +848,72 @@ glm::mat4 XrComponent::GetProjectionMatrixForCurrentViewAndSetFrustumPlanes(floa
...
+    {
+        const XrFovf& fov = mView.fov;
+
+        const float left  = tanf(fov.angleLeft) * mNearClip;
+        const float right = tanf(fov.angleRight) * mNearClip;
+
+        const float down = tanf(fov.angleDown) * mNearClip;
+        const float up   = tanf(fov.angleUp) * mNearClip;
+
+        // Note: the current variant in use is frustumRH_ZO.
```
Comment (Keenuts) [inline src/ppx/xr_component.cpp:903]: super-duper-nit: this kind of comment is bound to become invalid at some point since changing global constants is done somewhere else.

**Review by Keenuts — APPROVED** [review]

File: src/ppx/xr_component.cpp (hunk 13 above)
Comment (hysw) [reply src/ppx/xr_component.cpp:903]: It's somewhat intentional, I wrote down this comment so in case something changed in the future and we discover xr need to stay on the same  `glm::frustum` variant, we know which one it is.

File: projects/04_cube_xr/CubeXr.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (hysw) [reply projects/04_cube_xr/CubeXr.cpp:278]: Okay, I'll delete that line since `ppx::Camera` without modification initialize to `PPX_CAMERA_DEFAULT_{NEAR,FAR}_CLIP`
